returnn/forward.py

```diff
@@ -0,0 +1,201 @@
+            device=self.device,
+        )
+        config.write(self.out_returnn_config_file.get_path())
+
+        cmd = [
+            self.returnn_python_exe.get_path(),
+            os.path.join(self.returnn_root.get_path(), "rnn.py"),
+            self.out_returnn_config_file.get_path(),
+        ]
+        util.create_executable("rnn.sh", cmd)
+
+        # check here if model actually exists
+        assert os.path.exists(
+            self.model_checkpoint.index_path.get_path()
+        ), "Provided model does not exists: %s" % str(self.model_checkpoint)
+
+    def run(self):
+        # run everything in a TempDir as writing HDFs can cause heavy load
+        with tempfile.TemporaryDirectory(prefix=gs.TMP_PREFIX) as d:
+            print("using temp-dir: %s" % d)
+            call = [
+                self.returnn_python_exe.get_path(),
+                os.path.join(self.returnn_root.get_path(), "rnn.py"),
+                self.out_returnn_config_file.get_path(),
+            ]
+
+            try:
+                sp.check_call(call, cwd=d)
+            except Exception as e:
+                print("Run crashed - copy temporary work folder as 'crash_dir'")
+                shutil.copytree(d, "crash_dir")
+                raise e
+
+            # move log and tensorboard
+            shutil.move(os.path.join(d, "returnn.log"), "returnn.log")
+            tensorboard_dirs = glob.glob(os.path.join(d, "eval-*"))
+            for dir in tensorboard_dirs:
+                shutil.move(dir, os.path.basename(dir))
+
+            # move hdf outputs to output folder
+            for k, v in self.out_hdf_files.items():
+                shutil.move(os.path.join(d, k), v.get_path())
+
+    @classmethod
+    def create_returnn_config(
+        cls,
+        model_checkpoint: Checkpoint,
+        returnn_config: ReturnnConfig,
+        eval_mode: bool,
+        log_verbosity: int,
+        device: str,
+        **kwargs,
+    ):
+        """
+        Update the config locally to make it ready for the forward/eval task.
+        The resulting config will be used for hashing.
+
+        :param model_checkpoint:
+        :param returnn_config:
+        :param log_verbosity:
+        :param device:
+        :param kwargs:
+        :return:
+        """
+        assert device in ["gpu", "cpu"]
+        assert "task" not in returnn_config.config
+        assert "load" not in returnn_config.config
+        assert "model" not in returnn_config.config
+
+        res = copy.deepcopy(returnn_config)
+
+        config = {
+            "load": model_checkpoint,
+            "task": "eval" if eval_mode else "forward",
+        }
+
+        post_config = {
+            "device": device,
+            "log": ["./returnn.log"],
+            "log_verbosity": log_verbosity,
+        }
+
+        if not eval_mode:
+            post_config["forward_override_hdf_output"] = True
+            post_config["output_file"] = "output.hdf"
```

Comment:
didnt this cause problems with the file server?? because we are writing directly to network storage.